File: plugins/templates/AudioFilterModule.m
#import "AudioFilterModule.h"
#import <React/RCTLog.h>
#import <AVFoundation/AVFoundation.h>

@implementation AudioFilterModule {
  AVAudioEngine *_audioEngine;
  AVAudioPlayerNode *_playerNode;
  AVAudioUnitEQ *_eqUnit;
}

// Expose this module to JS under the name “AudioFilterModule”
RCT_EXPORT_MODULE();

// JS method: filterAndPlay(localFilePath: string, cutoffFreq: number)
RCT_EXPORT_METHOD(filterAndPlay:(NSString *)filePath cutoffFreq:(nonnull NSNumber *)cutoffFreq)
{
  // Lazy-init audio engine
  if (!_audioEngine) {
    _audioEngine = [[AVAudioEngine alloc] init];
    _playerNode = [[AVAudioPlayerNode alloc] init];
    _eqUnit = [[AVAudioUnitEQ alloc] initWithNumberOfBands:1];

    AVAudioUnitEQFilterParameters *band = _eqUnit.bands.firstObject;
    band.filterType = AVAudioUnitEQFilterTypeLowPass;
    band.frequency = cutoffFreq.floatValue;
    band.bypass = NO;

    [_audioEngine attachNode:_playerNode];
    [_audioEngine attachNode:_eqUnit];
    [_audioEngine connect:_playerNode to:_eqUnit format:nil];
    [_audioEngine connect:_eqUnit to:_audioEngine.mainMixerNode format:nil];

    NSError *err = nil;
    [_audioEngine startAndReturnError:&err];
    if (err) {
      RCTLogError(@"AudioEngine failed to start: %@", err.localizedDescription);
      return;
    }
  }

  // Play the file through the low-pass filter
  NSURL *url = [NSURL fileURLWithPath:filePath];
  AVAudioFile *audioFile = [[AVAudioFile alloc] initForReading:url error:nil];
  [_playerNode scheduleFile:audioFile atTime:nil completionHandler:nil];
  [_playerNode play];
}

@end

File: plugins/templates/AudioFilterModule.swift
import Foundation
import AVFoundation

@objc(AudioFilterModule)
class AudioFilterModule: NSObject {
  private var engine: AVAudioEngine?
  private var playerNode: AVAudioPlayerNode?
  private var eq: AVAudioUnitEQ?

  @objc
  func filterAndPlay(_ filePath: NSString, cutoff: NSNumber) {
    let url = URL(fileURLWithPath: filePath as String)
    engine     = AVAudioEngine()
    playerNode = AVAudioPlayerNode()
    eq         = AVAudioUnitEQ(numberOfBands: 1)

    guard let engine = engine,
          let player = playerNode,
          let eq     = eq else { return }

    let band = eq.bands[0]
    band.filterType = .lowPass
    band.frequency  = cutoff.floatValue
    band.bypass     = false

    engine.attach(player)
    engine.attach(eq)
    engine.connect(player, to: eq, format: nil)
    engine.connect(eq, to: engine.mainMixerNode, format: nil)

    do {
      let audioFile = try AVAudioFile(forReading: url)
      player.scheduleFile(audioFile, at: nil)
      try engine.start()
      player.play()
    } catch {
      print("AudioFilterModule error:", error)
    }
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    true
  }
}
